File: ai/on_device_cv/contracts.py
"""Backend-neutral data contracts used by the on-device CV pipeline."""

from __future__ import annotations

import math
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

import numpy as np


@dataclass(frozen=True, slots=True)
class BoundingBox:
    """A normalized ``xyxy`` box in the original upright input frame."""

    x_min: float
    y_min: float
    x_max: float
    y_max: float

    def __post_init__(self) -> None:
        coordinates = (self.x_min, self.y_min, self.x_max, self.y_max)
        if not all(math.isfinite(value) for value in coordinates):
            raise ValueError("Bounding-box coordinates must be finite")
        if not all(0.0 <= value <= 1.0 for value in coordinates):
            raise ValueError(f"Bounding box must be normalized to [0, 1]: {coordinates}")
        if self.x_min >= self.x_max or self.y_min >= self.y_max:
            raise ValueError(f"Bounding box must have positive area: {coordinates}")

    @classmethod
    def clipped(
        cls,
        x_min: float,
        y_min: float,
        x_max: float,
        y_max: float,
    ) -> BoundingBox | None:
        """Clip raw coordinates to the contract, dropping invalid boxes."""

        coordinates = (x_min, y_min, x_max, y_max)
        if not all(math.isfinite(value) for value in coordinates):
            return None

        clipped = tuple(min(1.0, max(0.0, float(value))) for value in coordinates)
        if clipped[0] >= clipped[2] or clipped[1] >= clipped[3]:
            return None
        return cls(*clipped)

    def to_pixels(self, width: int, height: int) -> tuple[int, int, int, int]:
        """Convert the normalized box to pixel coordinates for display."""

        if width <= 0 or height <= 0:
            raise ValueError("Frame width and height must be positive")
        return (
            round(self.x_min * width),
            round(self.y_min * height),
            round(self.x_max * width),
            round(self.y_max * height),
        )

    def iou(self, other: BoundingBox) -> float:
        """Return intersection-over-union with another normalized box."""

        intersection_width = max(0.0, min(self.x_max, other.x_max) - max(self.x_min, other.x_min))
        intersection_height = max(0.0, min(self.y_max, other.y_max) - max(self.y_min, other.y_min))
        intersection = intersection_width * intersection_height
        own_area = (self.x_max - self.x_min) * (self.y_max - self.y_min)
        other_area = (other.x_max - other.x_min) * (other.y_max - other.y_min)
        union = own_area + other_area - intersection
        return intersection / union if union > 0.0 else 0.0

    def to_dict(self) -> dict[str, float]:
        return {
            "x_min": float(self.x_min),
            "y_min": float(self.y_min),
            "x_max": float(self.x_max),
            "y_max": float(self.y_max),
        }


@dataclass(frozen=True, slots=True)
class DetectionResult:
    """Post-processed detector output independent of model tensor formats."""

    label: str
    confidence: float
    bbox: BoundingBox
    class_id: int | None = None

    def __post_init__(self) -> None:
        if not self.label or not self.label.strip():
            raise ValueError("Detection label must not be empty")
        if not math.isfinite(self.confidence) or not 0.0 <= self.confidence <= 1.0:
            raise ValueError(f"Detection confidence must be in [0, 1]: {self.confidence}")
        if self.class_id is not None and (type(self.class_id) is not int or self.class_id < 0):
            raise ValueError("class_id must be a non-negative integer when provided")


@dataclass(frozen=True, slots=True)
class TrackedObject:
    """A currently observed detection with a stream-persistent track ID."""

    track_id: int
    label: str
    confidence: float
    bbox: BoundingBox

    def __post_init__(self) -> None:
        if type(self.track_id) is not int or self.track_id <= 0:
            raise ValueError("track_id must be a positive integer")
        if not self.label or not self.label.strip():
            raise ValueError("Tracked-object label must not be empty")
        if not math.isfinite(self.confidence) or not 0.0 <= self.confidence <= 1.0:
            raise ValueError(f"Tracked-object confidence must be in [0, 1]: {self.confidence}")

    def to_dict(self) -> dict[str, Any]:
        return {
            "track_id": self.track_id,
            "label": self.label,
            "confidence": float(self.confidence),
            "bbox": self.bbox.to_dict(),
        }


@dataclass(frozen=True, slots=True)
class FrameResult:
    """Public per-frame response returned by :class:`OnDeviceCVPipeline`."""

    objects: tuple[TrackedObject, ...] = ()

    def __post_init__(self) -> None:
        track_ids = [item.track_id for item in self.objects]
        if len(track_ids) != len(set(track_ids)):
            raise ValueError("A frame cannot contain duplicate track IDs")

    @classmethod
    def from_objects(cls, objects: Iterable[TrackedObject]) -> FrameResult:
        return cls(tuple(objects))

    def to_dict(self) -> dict[str, list[dict[str, Any]]]:
        """Serialize to the stable Snap-Sight API schema."""

        return {"objects": [item.to_dict() for item in self.objects]}


def validate_bgr_frame(frame_bgr: np.ndarray) -> None:
    """Validate the single frame format shared by all PC detector adapters."""

    if not isinstance(frame_bgr, np.ndarray):
        raise TypeError("frame must be a numpy array")
    if frame_bgr.dtype != np.uint8:
        raise ValueError(f"frame dtype must be uint8, got {frame_bgr.dtype}")
    if frame_bgr.ndim != 3 or frame_bgr.shape[2] != 3:
        raise ValueError(f"frame shape must be HxWx3, got {frame_bgr.shape}")
    if frame_bgr.shape[0] <= 0 or frame_bgr.shape[1] <= 0:
        raise ValueError("frame must not be empty")

File: frontend/app/src/main/java/com/example/snap_sight/cv/Contracts.kt
package com.example.snap_sight.cv

import java.util.Locale

/**
 * [② 온디바이스 CV] `ai/on_device_cv/contracts.py` 의 Kotlin 포팅.
 *
 * Python PC 프로토타입과 **같은 공개 계약**을 유지한다. 값 의미가 달라지면
 * 양쪽을 함께 고쳐야 하며, JSON 키 이름은 백엔드/테스트가 의존하므로 바꾸지 않는다.
 *
 * 이 파일은 android.* 에 의존하지 않는다 → JVM 단위 테스트(`src/test`)에서 그대로 검증 가능.
 */

/** normalized `xyxy` 박스. 원본 upright 프레임 기준, 좌측 상단 원점, 모든 좌표 `[0, 1]`. */
data class BoundingBox(
    val xMin: Float,
    val yMin: Float,
    val xMax: Float,
    val yMax: Float,
) {
    init {
        require(xMin.isFinite() && yMin.isFinite() && xMax.isFinite() && yMax.isFinite()) {
            "Bounding-box coordinates must be finite"
        }
        require(xMin in 0f..1f && yMin in 0f..1f && xMax in 0f..1f && yMax in 0f..1f) {
            "Bounding box must be normalized to [0, 1]: ($xMin, $yMin, $xMax, $yMax)"
        }
        require(xMin < xMax && yMin < yMax) {
            "Bounding box must have positive area: ($xMin, $yMin, $xMax, $yMax)"
        }
    }

    val width: Float get() = xMax - xMin
    val height: Float get() = yMax - yMin

    /** 편차 계산(단계 5)이 바로 쓸 수 있는 기하 값. 여기서 정책 판단은 하지 않는다. */
    val centerX: Float get() = (xMin + xMax) / 2f
    val centerY: Float get() = (yMin + yMax) / 2f
    val area: Float get() = width * height

    fun iou(other: BoundingBox): Float {
        val intersectionWidth = (minOf(xMax, other.xMax) - maxOf(xMin, other.xMin)).coerceAtLeast(0f)
        val intersectionHeight =
            (minOf(yMax, other.yMax) - maxOf(yMin, other.yMin)).coerceAtLeast(0f)
        val intersection = intersectionWidth * intersectionHeight
        val union = area + other.area - intersection
        return if (union > 0f) intersection / union else 0f
    }

    /** ⑥ 오버레이/포커스 좌표용 픽셀 변환. */
    fun toPixels(frameWidth: Int, frameHeight: Int): PixelRect {
        require(frameWidth > 0 && frameHeight > 0) { "Frame width and height must be positive" }
        return PixelRect(
            left = Math.round(xMin * frameWidth),
            top = Math.round(yMin * frameHeight),
            right = Math.round(xMax * frameWidth),
            bottom = Math.round(yMax * frameHeight),
        )
    }

    fun toJson(): String = buildString {
        append("{\"x_min\":").append(coordinate(xMin))
        append(",\"y_min\":").append(coordinate(yMin))
        append(",\"x_max\":").append(coordinate(xMax))
        append(",\"y_max\":").append(coordinate(yMax))
        append('}')
    }

    companion object {
        /**
         * 원시 좌표를 계약 범위로 clip 한다. 면적이 0 이하가 되거나 비정상 값이면 null.
         * detector adapter 가 모델 출력을 정리할 때 쓰는 진입점이다.
         */
        fun clipped(xMin: Float, yMin: Float, xMax: Float, yMax: Float): BoundingBox? {
            if (!xMin.isFinite() || !yMin.isFinite() || !xMax.isFinite() || !yMax.isFinite()) {
                return null
            }
            val left = xMin.coerceIn(0f, 1f)
            val top = yMin.coerceIn(0f, 1f)
            val right = xMax.coerceIn(0f, 1f)
            val bottom = yMax.coerceIn(0f, 1f)
            if (left >= right || top >= bottom) return null
            return BoundingBox(left, top, right, bottom)
        }
    }
}

data class PixelRect(val left: Int, val top: Int, val right: Int, val bottom: Int)

/** detector 가 후처리를 끝낸 결과. 모델 텐서 포맷과 무관해야 한다. */
data class Detection(
    val label: String,
    val confidence: Float,
    val bbox: BoundingBox,
    val classId: Int? = null,
) {
    init {
        require(label.isNotBlank()) { "Detection label must not be empty" }
        require(confidence.isFinite() && confidence in 0f..1f) {
            "Detection confidence must be in [0, 1]: $confidence"
        }
        require(classId == null || classId >= 0) {
            "classId must be a non-negative integer when provided"
        }
    }
}

/** 스트림 내내 유지되는 `track_id` 가 붙은, 현재 프레임에서 실제로 관측된 객체. */
data class TrackedObject(
    val trackId: Int,
    val label: String,
    val confidence: Float,
    val bbox: BoundingBox,
    val classId: Int? = null,
) {
    init {
        require(trackId > 0) { "trackId must be a positive integer" }
        require(label.isNotBlank()) { "Tracked-object label must not be empty" }
        require(confidence.isFinite() && confidence in 0f..1f) {
            "Tracked-object confidence must be in [0, 1]: $confidence"
        }
        require(classId == null || classId >= 0) {
            "classId must be a non-negative integer when provided"
        }
    }

    fun toJson(): String = buildString {
        append("{\"track_id\":").append(trackId)
        append(",\"label\":\"").append(escapeJson(label)).append('"')
        append(",\"confidence\":").append(score(confidence))
        append(",\"bbox\":").append(bbox.toJson())
        append('}')
    }
}

/**
 * 프레임 1장의 공개 응답. **이게 ⑤/③ 로 넘기는 계약이다.**
 *
 * ```json
 * {"objects":[{"track_id":17,"label":"person","confidence":0.94,
 *              "bbox":{"x_min":0.31,"y_min":0.12,"x_max":0.68,"y_max":0.91}}]}
 * ```
 */
data class FrameResult(val objects: List<TrackedObject> = emptyList()) {
    init {
        require(objects.map { it.trackId }.toSet().size == objects.size) {
            "A frame cannot contain duplicate track IDs"
        }
    }

    val isEmpty: Boolean get() = objects.isEmpty()

    fun toJson(): String = objects.joinToString(
        separator = ",",
        prefix = "{\"objects\":[",
        postfix = "]}",
    ) { it.toJson() }

    companion object {
        val EMPTY = FrameResult()

        fun fromObjects(objects: Iterable<TrackedObject>): FrameResult =
            FrameResult(objects.sortedBy { it.trackId })
    }
}

/** bbox 는 6자리, confidence 는 4자리로 고정 직렬화한다 (로그 diff 안정성). */
private fun coordinate(value: Float): String = String.format(Locale.US, "%.6f", value)

private fun score(value: Float): String = String.format(Locale.US, "%.4f", value)

/** Objects365 label 은 `cabinet/shelf` 처럼 슬래시를 포함한다. 최소 escape 만 수행. */
private fun escapeJson(value: String): String {
    if (value.none { it == '"' || it == '\\' || it.code < 0x20 }) return value
    return buildString(value.length + 8) {
        for (character in value) {
            when {
                character == '"' -> append("\\\"")
                character == '\\' -> append("\\\\")
                character == '\n' -> append("\\n")
                character == '\r' -> append("\\r")
                character == '\t' -> append("\\t")
                character.code < 0x20 -> append(String.format(Locale.US, "\\u%04x", character.code))
                else -> append(character)
            }
        }
    }
}
